pkg/visitors/assemble.go

```diff
@@ -71,6 +72,86 @@ func (v *Assemble) Visit(f uefi.Firmware) error {
 		}
 
 		fileOffset := f.DataOffset
+
+		var pad *uefi.File
+		var shrink uint64
+		for _, file := range f.Files {
+			fileBuf := file.Buf()
+			fileLen := uint64(len(fileBuf))
+			if fileLen == 0 {
+				log.Fatal("Unexpected empty file buffer (at least a header was expected): %v", file.Header.GUID)
+			}
+			if file.Header.GUID == *uefi.FFGUID && fileLen > 24 {
+				// found non-empty PAD to check for free space.
+				pad = file
+			}
+
+			// Pad to the 8 byte alignments.
+			alignedOffset := uefi.Align8(fileOffset)
+			// Read out the file alignment requirements
+			if alignBase := file.Header.Attributes.GetAlignment(); alignBase != 1 {
+				hl := file.HeaderLen()
+				// We need to align the data, not the header. This is so terrible.
+				fileDataOffset := uefi.Align(alignedOffset+hl, alignBase)
+				// Calculate the starting offset of the file
+				newOffset := fileDataOffset - hl
+				if gap := (newOffset - alignedOffset); gap >= 8 && gap < uefi.FileHeaderMinLength {
+					// We need to re align to the next boundary cause we can't put a pad file in here.
+					// Who thought this was a good idea?
+					fileDataOffset = uefi.Align(fileDataOffset+1, alignBase)
+					newOffset = fileDataOffset - hl
+				}
+				if newOffset != alignedOffset {
+					// Add a pad file starting from alignedOffset to newOffset
+					alignedOffset = newOffset
+				}
+			}
+			fileOffset = alignedOffset + fileLen
+		}
+
+		// Check if we're out of space.
+		if f.Length < fileOffset && !f.Resizable {
+			// do we have a non-empty pad file?
+			if pad != nil {
+				shrink = fileOffset - f.Length
+				trim := shrink
+				if trim > (uint64(len(pad.Buf()) + 24)) {
+					return fmt.Errorf("out of space in non-empty padding. space available: %v bytes, need apace: %v bytes", len(pad.Buf()), trim)
+				}
+				// remove free space after the header from pad file and recalculate
+				tmp_header := pad.Buf()[:24]
```

Comment:
Are you trying to create a file or section header here? There are data structures in pkg/uefi/file.go and pkg/uefi/section.go that can create that for you, Please try not to write the offsets manually by addressing the bytes.

Reply:
It's a section header.  I'll look at that.  I thought I asked about how to do this a better way before.  I'd rather not do it this way but didn't see a different way.